ai8x.py

```diff
@@ -563,6 +572,148 @@ def get_activation(activation=None):
     return Empty()
 
 
+def histogram(inp, bins):
+    """
+    CUDA compatible histogram calculation
+    """
+    minimum, maximum = inp.min(), inp.max()
+    counts = torch.histc(inp, bins, min=minimum, max=maximum).cpu()
+    boundaries = torch.linspace(minimum, maximum, bins + 1)
+    return counts, boundaries
+
+
+def calc_q_error(module, threshold, bits, eps=1e-9):
+    """
+    Activation quantization error calculation
+    """
+    quantized_hist = module.hist[1].clone()
+    quantized_hist = torch.round((quantized_hist / (threshold + eps)) * 2**(bits-1))
+    quantized_hist = torch.clamp(quantized_hist, -2**(bits-1), 2**(bits-1)-1)
+    quantized_hist = (quantized_hist * (threshold + eps) / 2**(bits-1))
+    err = torch.sum(((quantized_hist - module.hist[1])**2)*module.hist[0]) \
+        / torch.sum(module.hist[0])
+
+    return err
+
+
+def _merge_hist(module):
+    """
+    Merge histograms of activations
+    """
+    bins_to_stack = []
+    for hist in module.hist:
+        bins_to_stack.append(hist[1])
+    stacked_bins = torch.stack(bins_to_stack)
+    min_edge = stacked_bins.min()
+    max_edge = stacked_bins.max()
+    # 2048 is the number of bins
+    width = (max_edge - min_edge) / 2048
+    merged_bins = torch.arange(min_edge.item(), (max_edge+width).item(), width.item())
```

Comment:
Depending on the exact value of `max_edge` (and `min_edge`), the shape of `merged_bins` can be 2049 or 2050. (the shape of 'interp_counts` reflects this number, so the rest of the algorithm doesn't throw any errors)

I would suggest using `torch.linspace` instead of `torch.arange` to ensure a fixed shape.

Reply:
Thanks, I addressed this in the latest commit.